src/charm.py

```diff
@@ -170,6 +191,48 @@ def _configure_charm_authorization(self):
             login_details_secret = self.model.get_secret(label=GOCERT_LOGIN_SECRET_LABEL)
             login_details_secret.set_content(login_details.to_dict())
 
+    def _configure_certificate_requirers(self):
+        """Get all CSR's and certs from databags and Gocert, compare differences and update requirers if needed."""
+        login_details = self._get_or_create_admin_account()
+        if not login_details or not login_details.token:
+            logger.warning("couldn't distribute certificates: not logged in")
+            return
+        databag_csrs = self.tls.get_certificate_requests()
+        gocert_csrs_table = self.client.get_certificate_requests_table(login_details.token)
+        if not gocert_csrs_table:
+            logger.warning("couldn't distribute certificates: couldn't get table from GoCert")
+            return
+        for request in databag_csrs:
+            matching_rows = list(
+                filter(
+                    lambda x: x.csr == str(request.certificate_signing_request),
+                    gocert_csrs_table.rows,
+                )
+            )
+            if len(matching_rows) < 1:
+                self.client.post_csr(str(request.certificate_signing_request), login_details.token)
+                continue
+            request_row = matching_rows[0]
+            provided_certificates = self.tls.get_issued_certificates(request.relation_id)
+            if request_row.certificate_chain == "rejected" or request_row.certificate_chain == "":
+                # TODO: tls lib v4 lacks the ability to revoke the certificate of a specific request
+                pass
+            certificate_chain = [
+                Certificate.from_string(cert) for cert in request_row.certificate_chain
+            ]
+            if len(certificate_chain) > 0 and certificate_chain[0] not in [
+                obj.certificate for obj in provided_certificates
```

Comment:
That's a good idea, I was confused since there is no function to just get a certificate given a csr, but it's possible to do it by looking at the ProviderCertificate object's `certificate_signing_request` field.

Reply:
It seems to me that we still are comparing the current certificate and the provided certificate. Is this intended? Do we want to allow providing a new certificate for the same CSR without any changes?

Reply:
Yes. We want to update the certificate in the databag if the certificate in gocert has changed.